rfcs/20190305-modular-tensorflow.md

```diff
@@ -0,0 +1,481 @@
+
+
+## Detailed Design
+
+We will describe each key design element here in detail. To make the points clearer, trivial examples will be created.
+
+### Modularity in C/C++
+
+This section will describe the key design points for C++ modularity.
+
+#### Modules
+
+Each module's main pieces will be a module interface in C, and a registry for plugins implementing this module. As a supporting piece, each module will also need to provide a mechanism for plugins to add themselves to the registry at runtime. Below is a toy example for the described design:
+
+
+```
+// A simple toy module called "M"
+typedef struct M_context;
+
+// The summary of the interface plugins need to implement
+typedef struct MInterface {
+  void (*M_Init)(M_context*, string);
+  void (*M_Execute)(M_context*);
+};
+
+// The plugin registry for module M
+// The registry here implemented uses C++ class std::map
+// This is OK as it is not exposed in our example here.
+// As far as module implementations are concerned, they only
+// need to see and use M_RegisterPlugin method, and that method takes // care of their addition into this registry.
+std::map<string, MInterface> m_plugin_registry;
+
+// Function to call for each plugin at load time
+// to add themselves to the registry
+void M_RegisterPlugin(
+    string plugin, 
+    void (*M_Init)(M_context*, string),
+    void (*M_Execute)(M_context*)) {
+  // Omitting error handling.
+  m_plugin_registry.insert(plugin,
+                           MInterface(M_Init, M_Execute));
+}
+
+// Implementation of the interface is just a thin layer to
+// get the correct plugin and call it.
+// Here we assume that plugin explicitly gets selected by
+// the init function. Some modules can go with implicit selection,
+// Such as deducing the filesystem from the file path.
+void M_Init(M_context* ctx, string id) {
+  // Just a quick hacky way to select the plugin here
+  ctx->interface = m_plugin_registry[id];
+  ctx->interface.M_Init(ctx, id);
+}
+
+void M_Execute(M_context* ctx) {
+  ctx->interface.M_Execute(ctx);
+}
+```
+
+
+**Please note that the above is a toy example to explain the abstract ideas described above. Exact implementation can vary across different modules.**
+
+Interface has to be pure C at the ABI level. We can have C++ header-only libraries built on top of these C ABI/APIs.
+
+
+#### Plugins
+
+Plugins need to include implementation of the interfaces declared by one module. If the module interface requires Init and Compute methods, it will need to implement these two functions, plus a TF_InitPlugin function which will be called at load time. This function will also need to register the plugin as prescribed by the module.
+
```

Comment:
SDK/API architecture has the potential to influence performance of plugins for better and worse. If an API design favors one plugin and hurts another, what is the plan to resolve the conflict?

Reply:
This is a great question. From our side, we will try to be as neutral as possible. Initial APIs will likely be based on the current C++ APIs TF has. But the API designs will be individually reviewed through different RFCs. We hope that we will be able to evaluate all concerns during these reviews. If we make mistakes favoring one or the other framework, we hope to alleviate these through addition of new symbols, and complete redesigns with the next major version releases.